Feature: uploading/view signature

ziweiwu opened the pull request:
Feature
- allow uploading a signature image file on create/update user page
- add a view button to view the signature image for the react home page and thymeleaf home page
- after click view button, a view signature page will display to show user signature 

Notes
- set function for signature field takes an input file uploaded in the form, convert it to a byte stream and store it in the database 
- `getEncodedSignature()` decode the byte[] signature data from the database into a string using Base64 encoding, it allows the string to be embedded in the HTML and rendered as a picture by browser

ziweiwu committed:
updated react user page to show view signature link

ziweiwu committed:
- added setSignature method to read signature from a file
- added a getEncodedSignature method to get a base64 encoded string from byte[] signature
- added a toString() method for User entity, can be handy for debugging purpose

ziweiwu committed:
added a Get mapping to view signature page

ziweiwu committed:
add view signature page for viewing user's signature picture

ziweiwu committed:
- allow user to upload a signature image file in jpg in create user / update user page
- added a view button in user list to view signature

ziweiwu committed:
added exception handling for setSignature

ziweiwu committed:
added password input validation in create/update user forms

anderMatt commented on src/main/java/github/paz/awardportal/model/User/User.java:
```diff
@@ -85,4 +89,51 @@ public void updateUser(BaseUser update){
+    }
+
+    // return the signature in base64 string, allow it to be displayed in html
+    public String getEncodedSignature() {
```
Looks good. However, what do you think about this: https://www.baeldung.com/spring-boot-jsoncomponent

We can customize how User objects are serialized to JSON such that the signature field, if present, is automatically converted to a Base64 string. That way, the front end just needs to access this field <img src="data:image/jpg;base64, {{user.signature}} />.

ziweiwu replied:
Looks like JSON outputs encoded string by default in `encodedSignature` field.

anderMatt commented on src/main/resources/templates/users/view-signature.html:
```diff
@@ -0,0 +1,23 @@
+
+<body>
+<p th:text="'User: ' +  ${user.firstName} + ' '+ ${user.lastName}"></p>
+<img th:src="@{'data:image/jpg;base64,'+${user.getEncodedSignature()}}"/>
```
See comment above - this works just fine if we want to mix Thymeleaf templates and our SPA frontend.

anderMatt commented on src/main/java/github/paz/awardportal/model/User/User.java:
```diff
@@ -85,4 +89,51 @@ public void updateUser(BaseUser update){
+
+    // return the signature in base64 string, allow it to be displayed in html
+    public String getEncodedSignature() {
+        try {
```
Java 8 introduced standard library support for this, which could simplify the code.

String encodedSignature = Base64.getEncoder().encodeToString(signature);

Just pointing this out, this way works perfectly fine too.

anderMatt replied:
import java.util.Base64

is the import statement

anderMatt reviewed (approved):
Looks good. I had two suggestions, up to you if you want to try them out or keep everything as is:

* Customize User serialization to JSON so that the byte[] signature field is automatically converted to a Base64 string

* Use the standard library java.util.Base64 class for converting the signature byte[] field to an encoded string

ziweiwu committed:
- fixed redirection of office controller
- added more strict input validation on the create/update forms for user and office
